pkg/ddl/executor.go

```diff
@@ -888,32 +888,83 @@ func GetDefaultCollation(cs string, defaultUTF8MB4Collation string) (string, err
 // ResolveCharsetCollation will resolve the charset and collate by the order of parameters:
 // * If any given ast.CharsetOpt is not empty, the resolved charset and collate will be returned.
 // * If all ast.CharsetOpts are empty, the default charset and collate will be returned.
+// * This is used for:
+//   - CREATE TABLE
+//   - CREATE SCHEMA (aka CREATE DATABASE)
+//   - ALTER TABLE ... ADD COLUMN
+//   - ALTER TABLE ... MODIFY COLUMN
 func ResolveCharsetCollation(charsetOpts []ast.CharsetOpt, utf8MB4DefaultColl string) (chs string, coll string, err error) {
-	for _, v := range charsetOpts {
-		if v.Col != "" {
-			collation, err := collate.GetCollationByName(v.Col)
+	// for CREATE TABLE:
+	//   objectChs: Table character set, if specified
+	//   objectColl: Table collation, if specified
+	//   tempChs: Schema character set
+	//   tempColl: Schema collation
+	// for CREATE SCHEMA:
+	//   objectChs: schema character set, if specified
+	//   objectColl: schema collation, if specified
+	//   tempChs, tempColl: empty
+	var templChs, templColl string
+	objectChs := charsetOpts[0].Chs
+	objectColl := charsetOpts[0].Col
+	if len(charsetOpts) > 1 {
+		templChs = charsetOpts[1].Chs
+		templColl = charsetOpts[1].Col
+	}
+
+	if objectChs != "" && objectColl != "" { // CHARACTER SET ... COLLATE ...
+		chs = objectChs
+		coll = objectColl
+	} else if objectChs != "" && objectColl == "" { // CHARACTER SET ...
+		chs = objectChs
+
+		if templColl != "" {
+			collation, err := collate.GetCollationByName(templColl)
 			if err != nil {
 				return "", "", errors.Trace(err)
 			}
-			if v.Chs != "" && collation.CharsetName != v.Chs {
-				return "", "", charset.ErrCollationCharsetMismatch.GenWithStackByArgs(v.Col, v.Chs)
+			if collation.CharsetName == chs {
+				coll = templColl
 			}
-			return collation.CharsetName, v.Col, nil
 		}
-		if v.Chs != "" {
-			coll, err := GetDefaultCollation(v.Chs, utf8MB4DefaultColl)
+		if coll == "" {
+			cs, err := charset.GetCharsetInfo(chs)
 			if err != nil {
 				return "", "", errors.Trace(err)
 			}
-			return v.Chs, coll, nil
+			coll = cs.DefaultCollation
 		}
+
+		if chs == charset.CharsetUTF8MB4 {
+			utf8mb4Coll := getDefaultCollationForUTF8MB4(chs, utf8MB4DefaultColl)
+			if utf8mb4Coll != "" {
+				coll = utf8mb4Coll
+			}
+		}
+	} else if objectChs == "" && objectColl != "" { // COLLATE ...
+		coll = objectColl
+		collation, err := collate.GetCollationByName(objectColl)
+		if err != nil {
+			return "", "", errors.Trace(err)
+		}
+		chs = collation.CharsetName
+	} else if objectChs == "" && objectColl == "" { // ...
+		chs = templChs
+		coll = templColl
 	}
-	chs, coll = charset.GetDefaultCharsetAndCollate()
-	utf8mb4Coll := getDefaultCollationForUTF8MB4(chs, utf8MB4DefaultColl)
-	if utf8mb4Coll != "" {
-		return chs, utf8mb4Coll, nil
+
+	if chs == "" || coll == "" {
```

Comment:
it's odd to have such a branch without handling cases that any one of it is empty, we can move above handling to here